persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/stringbased/JdbcStringBasedStore.java

```diff
@@ -531,11 +527,40 @@ public boolean contains(Object key) {
          }
          ps.setFetchSize(tableManager.getFetchSize());
          ResultSet rs = ps.executeQuery();
-         return function.apply(rs);
-      }, kvp -> {
-         JdbcUtil.safeClose(kvp.getValue());
-         connectionFactory.releaseConnection(kvp.getKey());
-      });
+         return function.apply(rs).doOnComplete(() -> JdbcUtil.safeClose(rs));
+      }, FlowableConnection::close);
+   }
+
+   class FlowableConnection {
+      final boolean autoCommit;
+      final ConnectionFactory factory;
+      final Connection connection;
+      final PreparedStatement statement;
+
+      FlowableConnection(ConnectionFactory factory, String sql) throws SQLException {
+         this.factory = factory;
+         this.connection = factory.getConnection();
+         this.autoCommit = connection.getAutoCommit();
+         this.statement = connection.prepareStatement(sql);
+
+         // Some JDBC drivers require auto commit disabled to do paging, however before calling setAutoCommit(false)
+         // we must ensure that we're not running in a managed transaction by ensuring that getAutoCommit is true.
+         // Without this check an exception would be thrown when calling setAutoCommit(false) during a managed transaction.
+         if (autoCommit)
+            connection.setAutoCommit(false);
+      }
+
+      void close() {
+         JdbcUtil.safeClose(statement);
+         if (autoCommit) {
+            try {
+               connection.rollback();
```

Comment:
I'm tempted to say that we should roll back before closing the statement, but I couldn't find a clear answer either way :)